### DISCUSSION TITLE
Do TypeBox support String Boolean type or Boolean Type Coercion?

### DISCUSSION THREAD

**Discussion by deanrih** [sinclairzx81/typebox#1067] [Q&A] [answered]
So I'm currently playing around with Zod and TypeBox for loading and parsing an env into a variable that can be used later (e.g. `env.ENV_NAME`), and I kinda like TypeBox more but there's some thing I'm still figuring out which is getting Boolean from a String or a type coercion, so in Zod, I can do something like

```ts
const schema = z.object({
  DEBUG: z.boolean({ coerce: true }),
});
```

and when given an env file of
```env
DEBUG="true"
```
the resulting of `env.DEBUG` is `boolean` and not a `string`

meanwhile doing it in TypeBox

```ts
const schema = Type.Object({
  DEBUG: Type.Boolean({ ... }),
});
```
there's no option to allow such behaviour, and using `Type.String` doesn't seem solve the problem since I need it to be `boolean`, furthermore I have additional behaviour to parsing logic for TypeBox, which roughly

- Get the env to a variable (something like `const currentEnv = process.env;`)
- Clean the env to remove any excess (`Value.Clean()`)
- Set the default env for those who have default option set (`Value.Default()`)
- Then check the env against the schema (`Value.Check()`)
- Before finally returned the decoded result (`Value.Decode()`)

Which is working fine when I have all env in string, but now I add `Type.Boolean` it fails the `Value.Check()` since the `DEBUG="true"` recognized as a `string` instead of `boolean` or not being coerced into `boolean`

Is it possible to do it in TypeBox or is it the limitation/out of the scope of this project?

**Comment by sinclairzx81** [comment, marked as answer]
@deanrih Hi,

The Value.Convert() function will coerce a string `"true"` into a boolean `true`. TypeBox treats this as an explicit call you need to make rather than via configuration on the type. The following loosely achieves this.

```typescript
const Environment = Type.Object({
  DEBUG: Type.Boolean()
})

const R = Value.Convert(Environment, process.env) // will convert "true" to true if possible.
```
However it's recommended you use Value.Parse for environment loading. TypeBox exposes quite a few low level Value operations for flexibility, but the Parse function internally implements all these operations in a standard way to process a value how you would expect.

```typescript
const Environment = Type.Object({
  DEBUG: Type.Boolean({ default: false }) // default values supported with Parse
})

const R = Value.Parse(Environment, process.env) // will convert "true" to true if possible.
```
I would have a go using Parse first. If you need custom environment variable processing, take a look at the TypeBox Parse implementation. This library encourages users to mix and match various components of the library to construct various pipelines for validation, the Parse function is a good place to start if you need that level of functionality.

https://github.com/sinclairzx81/typebox#values-parse

https://github.com/sinclairzx81/typebox/blob/master/src/value/parse/parse.ts#L44-L68

Hope this helps
S

**Reply by deanrih** [reply]
Ah, thank you, yeah I was expecting something with options and not explicit function call, but I tried and adding the `Value.Convert` just after `Value.Default` but before the `Value.Check` and it work

Thank you
